build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/SetCompatibleVersionsTask.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal.release;
+
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
+import com.github.javaparser.ast.expr.NameExpr;
+import com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter;
+
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+import org.gradle.initialization.layout.BuildLayout;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+public class SetCompatibleVersionsTask extends AbstractVersionsTask {
+
+    private Map<String, Integer> versionIds = Map.of();
+
+    @Inject
+    public SetCompatibleVersionsTask(BuildLayout layout) {
+        super(layout);
+    }
+
+    @Option(option = "version-id", description = "Version id used for the release. Of the form <VersionType>:<id>.")
+    public void versionIds(List<String> version) {
```

Comment:
This seems kind of odd. Why not have explicit options for each version type? As it is it's difficult to understand how I actually format this when there are multiple values. Are they comma separated?

Also, if you pass an unknown type it's simply ignored.

Reply:
By having the type as data it means we dont need to change the build automation to support more types, only the ES tasks. The build automation doesn't care about the version types used - see https://github.com/elastic/elasticsearch-infra/pull/375#discussion_r1457563114.

It allows 7.17 to emit `Version` only, which is then mapped properly by the 8.x tasks rather than the build automation. I've used the same mechanism here.

Reply:
These are specified the same as any other multi-value gradle option, by specifying the option & argument multiple times

Reply:
Got it, that way the release logic is branch agnostic. This makes sense.